django-stubs/contrib/admin/decorators.pyi

```diff
@@ -6,31 +6,33 @@ from django.contrib.admin.sites import AdminSite
 from django.db.models import Combinable, QuerySet
 from django.db.models.base import Model
 from django.db.models.expressions import BaseExpression
-from django.http import HttpRequest
+from django.http import HttpRequest, HttpResponseBase
 from django.utils.functional import _StrOrPromise
 from typing_extensions import TypeAlias
 
 _Model = TypeVar("_Model", bound=Model)
-_ModelAdmin = TypeVar("_ModelAdmin", bound=ModelAdmin)
+_ModelAdmin = TypeVar("_ModelAdmin", bound=ModelAdmin[Model])
 _Request = TypeVar("_Request", bound=HttpRequest)
-_QuerySet = TypeVar("_QuerySet", bound=QuerySet)
+_QuerySet = TypeVar("_QuerySet", bound=QuerySet[Model])
 # This is deliberately different from _DisplayT defined in contrib.admin.options
 _DisplayCallable: TypeAlias = Union[Callable[[_ModelAdmin, _Model], Any], Callable[[_Model], Any]]  # noqa: Y037
-_DisplayCallableT = TypeVar("_DisplayCallableT", bound=_DisplayCallable)
+_DisplayCallableT = TypeVar("_DisplayCallableT", bound=_DisplayCallable[ModelAdmin[Model], Model])
+_ActionReturn = TypeVar("_ActionReturn", bound=HttpResponseBase | None)
```

Comment:
Note: this currently builds off of my other branch, https://github.com/typeddjango/django-stubs/pull/1331.